File: backend-C#/DoctorsOffice/Services/VisitService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DoctorsOffice.DbModels;
using DoctorsOffice.Dto;
using DoctorsOffice.Repositories;

namespace DoctorsOffice.Services
{
    public class VisitService
    {
        private readonly VisitRepository _visitRepository;
        private readonly DoctorRepository _doctorRepository;
        private readonly PatientRepository _patientRepository;
        private readonly FacilityRepository _facilityRepository;

        public VisitService(VisitRepository visitRepository, FacilityRepository facilityRepository,
            PatientRepository patientRepository, DoctorRepository doctorRepository)
        {
            _visitRepository = visitRepository;
            _facilityRepository = facilityRepository;
            _patientRepository = patientRepository;
            _doctorRepository = doctorRepository;
        }

        private readonly List<string> _visitStartHours = new List<string>()
        {
            "8:00", "8:30",
            "9:00", "9:30",
            "10:00", "10:30",
            "11:00", "11:30",
            "12:00", "12:30",
            "13:00", "13:30",
            "14:00", "14:30",
            "15:00", "15:30"
        };

        private readonly List<string> _visitEndHours = new List<string>()
        {
            "8:30", "9:00",
            "9:30", "10:00",
            "10:30", "11:00",
            "11:30", "12:00",
            "12:30", "13:00",
            "13:30", "14:00",
            "14:30", "15:00",
            "15:30", "16:00"
        };

        public ICollection<Visit> GetVisitsForDoctor(string pwz, DateTime date)
        {
            return _visitRepository.FindAllByDoctorPwzAndDates(pwz, date.Date, date.Date.AddHours(23));
        }

        public List<string> CreateNoVisitHours()
        {
            return _visitStartHours
                .Select(visitStartHour =>
                {
                    var idx = _visitStartHours.IndexOf(visitStartHour);
                    return string.Concat(visitStartHour, "-", _visitEndHours[idx]);
                })
                .ToList();
        }

        public List<string> AvailableHoursFromGivenVisits(ICollection<Visit> visits)
        {
            var visitHours = visits.Select(visit =>
                    string
                        .Concat(visit.DateTime.Hour.ToString(), ":",
                            visit.DateTime.Minute == 0 ? "00" : visit.DateTime.Minute.ToString())
                )
                .ToList();

            return _visitStartHours
                .Select(visitStartHour =>
                {
                    var idx = _visitStartHours.IndexOf(visitStartHour);
                    return string.Concat(visitStartHour, "-", _visitEndHours[idx]);
                })
                .Where((visitHour => !visitHours.Contains(visitHour?.Split("-")[0])))
                .ToList();
        }

        public int CreateVisit(VisitCreate visitCreate)
        {
            return _visitRepository.Save(VisitCreateToVisit(visitCreate)).Id;
        }


        private Visit VisitCreateToVisit(VisitCreate visitCreate)
        {
            return new Visit
            {
                DateTime = visitCreate.DateTime, 
                Doctor = _doctorRepository.FindById(visitCreate.DoctorId),
                Patient = _patientRepository.Save(visitCreate.Patient),
                Facility = _facilityRepository.FindById(visitCreate.FacilityId)
            };
        }

        public List<Visit> GetVisitsForPatient(string identityNumber, string firstName, string lastName)
        {
            return _visitRepository.FindForPatient(identityNumber, firstName, lastName);
        }
    }
}

File: backend-Java/src/main/java/pl/kacste/visit/VisitService.java
package pl.kacste.visit;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.kacste.doctor.DoctorRepository;
import pl.kacste.facility.FacilityRepository;
import pl.kacste.patient.PatientRepository;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class VisitService {

    private final VisitRepository visitRepository;
    private final DoctorRepository doctorRepository;
    private final FacilityRepository facilityRepository;
    private final PatientRepository patientRepository;

    private final List<String> visitStartHours =
            List.of("8:00", "8:30",
                    "9:00", "9:30",
                    "10:00", "10:30",
                    "11:00", "11:30",
                    "12:00", "12:30",
                    "13:00", "13:30",
                    "14:00", "14:30",
                    "15:00", "15:30");
    private final List<String> visitEndHours =
            List.of("8:30", "9:00",
                    "9:30", "10:00",
                    "10:30", "11:00",
                    "11:30", "12:00",
                    "12:30", "13:00",
                    "13:30", "14:00",
                    "14:30", "15:00",
                    "15:30", "16:00");

    public Long createVisit(Visit visit){
       return visitRepository.save(visit).getId();
    }

    public List<Visit> getVisitsForDoctor(String pwz, LocalDate date) {
        return visitRepository.finaAllByDoctorPwzAndDates(
                pwz,
                LocalDateTime.of(date, LocalTime.MIDNIGHT),
                LocalDateTime.of(date, LocalTime.MIDNIGHT).plus(23, ChronoUnit.HOURS));
    }

    public List<String> availableHoursFromGivenVisits(List<Visit> visits){
        List<String> visitHours = visits
                .stream()
                .map(visit -> new StringBuilder(String.valueOf(visit.getDateTime().getHour()))
                        .append(":")
                        .append(visit.getDateTime().getMinute() == 0 ? "00" : visit.getDateTime().getMinute())
                        .toString())
                .collect(Collectors.toList());



        return visitStartHours
                .stream()
                .map((visitStartHour) -> {
                    int idx = visitStartHours.indexOf(visitStartHour);
                    return visitStartHour.concat("-").concat(visitEndHours.get(idx));
                })
                .filter(visitHourConcat -> !visitHours.contains(visitHourConcat.split("-")[0]))
                .collect(Collectors.toList());
    }

    public List<String> createNoVisitHours(){
        return visitStartHours
                .stream()
                .map(visitStartHour -> {
                    int idx = visitStartHours.indexOf(visitStartHour);
                    return visitStartHour.concat("-").concat(visitEndHours.get(idx));
                })
                .collect(Collectors.toList());
    }

    public Visit visitCreateToVisit(VisitCreateDTO visitCreate){
        return Visit
                .builder()
                .dateTime(visitCreate.getDateTime())
                .doctor(doctorRepository.getOne(visitCreate.getDoctorId()))
                .patient(patientRepository.save(visitCreate.getPatient()))
                .facility(facilityRepository.getOne(visitCreate.getFacilityId()))
                .build();

    }

    public List<Visit> getVisitsForPatient(String identityNumber, String firstName, String lastName){
        return visitRepository.findForPatient(identityNumber, firstName, lastName);
    }
}
